packages/server/app/Controllers/Http/AuthenticationController.ts

```diff
@@ -12,22 +15,96 @@ export const SignInRequestBody = z.object({
-        const userID = datatype.uuid();
-        const userNickname = internet.userName();
+    public async signUp({
+        request,
+        response,
+        auth,
+    }: HttpContextContract): Promise<SignUpResponseBody> {
+        try {
+            const { authenticationMode, email, password, userNickname } =
+                SignUpRequestBody.parse(request.body());
 
-        await User.create({
-            uuid: userID,
-            nickname: userNickname,
-            email: internet.email(),
-            password: internet.password(),
-        });
+            const userWithGivenNickname = await User.findBy(
+                'nickname',
+                userNickname,
+            );
 
-        return {
-            userID,
-            userNickname,
-        };
+            if (userWithGivenNickname) {
+                response.status(400);
+                return {
+                    status: 'UNAVAILABLE_NICKNAME',
+                };
+            }
+
+            const userWithGivenEmail = await User.findBy('email', email);
+            if (userWithGivenEmail) {
+                response.status(400);
+                return {
+                    status: 'UNAVAILABLE_EMAIL',
+                };
+            }
+
+            const passwordIsStrong = passwordStrengthRegex.test(password);
+            const passwordIsNotStrong = !passwordIsStrong;
+            if (passwordIsNotStrong) {
+                response.status(400);
+                return {
+                    status: 'WEAK_PASSWORD',
+                };
+            }
+
+            const { nickname, uuid: userID } = await User.create({
+                nickname: userNickname,
+                email: email,
+                password,
+            });
+
+            const userSummary: UserSummary = {
+                nickname,
+                userID,
+            };
+
+            switch (authenticationMode) {
+                case 'api': {
+                    const { token } = await auth
+                        .use('api')
+                        .attempt(email, password);
+
+                    return {
+                        token,
+                        userSummary,
+                        status: 'SUCCESS',
+                    };
+                }
+                case 'web': {
+                    await auth.use('web').attempt(email, password);
+
+                    return {
+                        userSummary,
+                        status: 'SUCCESS',
+                    };
+                }
+                default: {
+                    throw new Error(
+                        `unknown authentication mode encountered ${authenticationMode}`,
+                    );
+                }
+            }
+        } catch (e) {
+            console.log(e);
+
+            // if zod fails here we consider it as bad email edge case
+            // as the authenticationMode should not be settable by the user
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏I think we can assume an error would be thrown only because of an invalid email, but actually async operations such as `User.findBy` can fail, and ideally I think we should return a 500 response with a generic error status in this case.

We could determine specifically if the email is invalid by using `SignUpRequestBody.safeParse()`, which returns the error without throwing it, which is more convenient for such tests.

**I ask for no change**. These are just some thoughts.

Reply:
Indeed I think this is a great idea
Going deeper